chia/data_layer/s3_plugin_service.py

```diff
@@ -0,0 +1,164 @@
+
+    async def check_store_id(self, request: web.Request) -> web.Response:
+        self.update_instance_from_config()
+        try:
+            data = await request.json()
+        except Exception as e:
+            print(f"failed parsing request {request} {e}")
+            return web.json_response({"handles_url": False})
+        store_id = bytes32.from_hexstr(data["id"])
+        if store_id.hex() in self.store_ids:
+            return web.json_response({"handles_store": True})
+        return web.json_response({"handles_store": False})
+
+    async def upload(self, request: web.Request) -> web.Response:
+        try:
+            data = await request.json()
+            store_id = bytes32.from_hexstr(data["id"])
+            bucket = self.get_bucket(store_id)
+            try:
+                full_tree_path = Path(data["full_tree_path"])
+                diff_path = Path(data["diff_path"])
+                self.boto_client.upload_file(full_tree_path, bucket, full_tree_path.name)
+                self.boto_client.upload_file(diff_path, bucket, diff_path.name)
+            except ClientError as e:
+                print(f"failed uploading file to aws {e}")
+                return web.json_response({"uploaded": False})
+        except Exception as e:
+            print(f"failed handling request {request} {e}")
+            return web.json_response({"handles_url": False})
+        return web.json_response({"uploaded": True})
+
+    async def check_url(self, request: web.Request) -> web.Response:
+        self.update_instance_from_config()
+        try:
+            data = await request.json()
+        except Exception as e:
+            print(f"failed parsing request {request} {e}")
+            return web.json_response({"handles_url": False})
+        parse_result = urlparse(data["url"])
+        if parse_result.scheme == "s3" and data["url"] in self.urls:
+            return web.json_response({"handles_url": True})
+        return web.json_response({"handles_url": False})
+
+    async def download(self, request: web.Request) -> web.Response:
+        try:
+            data = await request.json()
+        except Exception as e:
+            print(f"failed parsing request {request} {e}")
+            return web.json_response({"handles_url": False})
+        url = data["url"]
+        client_folder = Path(data["client_folder"])
+        filename = data["filename"]
+        parse_result = urlparse(url)
+        bucket = parse_result.netloc
+        target_filename = client_folder.joinpath(filename)
+        # Create folder for parent directory
+        target_filename.parent.mkdir(parents=True, exist_ok=True)
+        with concurrent.futures.ThreadPoolExecutor() as pool:
+            await asyncio.get_running_loop().run_in_executor(
+                pool, functools.partial(self.boto_client.download_file, bucket, filename, str(target_filename))
+            )
+
+        return web.json_response({"downloaded": True})
```

Comment:
I'm not sure what pattern we want to put in a reference, but the lack of error handling with a controlled response here is inconsistent.